rfcs/text/1020-link-entities.md

```diff
@@ -0,0 +1,385 @@
+
+- Entity Type -> Link Type (constrains the types of links that can come from an entity)
+- Entity Type -> Entity Type (constrains the types of entities that can be linked to from an entity)
+- Entity Type -> Property Type (constrains the properties that can appear on an entity)
+- Property Type -> Property Type (constrains the types of properties that can appear within an object in another property)
+- Property Type -> Data Type (constrains the type of data that can be stored within a property)
+- Entity -> Entity Type (describes the associated entity type of an entity)
+- Entity -> Link (the outgoing or incoming link of an entity)
+- Link -> Entity (the source or target entity of a link)
+
+## Creating Higher-order abstractions through composition of the primitives of the type system
+
+### Groups
+
+Due to the compositional design of the primitives of the Type System, it's possible to create high-order abstractions within the system.
+For example, one could want to define "groups" of entities within their system, in such a way that they would be programmatically resolvable through blocks, _with constraints guaranteed by the specification_.
+
+One could currently do this by creating a "Group" entity type, and an associated "Has Element" link type.
+They could then create a specific "Group" entity, and link it to the entities they wish to be part of the group:
+
+```text
+       ┌─────┐
+       │Group│
+       └──┬──┘
+          │
+          │
+ ┌───Has element───┐
+ │     │     │     │
+ │     │     │     │
+ ▼     ▼     ▼     ▼
+┌─┐   ┌─┐   ┌─┐   ┌─┐
+│A│   │B│   │C│   │D│
+└─┘   └─┘   └─┘   └─┘
+```
+
+Being able to create such abstractions is a powerful feature of the Type System, as users are able to implicitly benefit from some of the following scenarios:
+
+- A user updates entity `A`, the link is still associated with the new version
+- A user deletes entity `A`, the Embedding Application (EA) knows that a link has been invalidated and is able to orchestrate/trigger a resolution flow
+- A user redefines the destination constraints of `Has element` links on the `Group` entity type to disallow entities of type `D`.
+  The EA can identify that the `D` entity breaks validation, and is able to orchestrate/trigger a resolution flow
+
+### Paths
+
+One might think a similar approach could be taken to define a "Path" entity type, also utilizing an associated "Has Element" link type.
+
+Take the following simple subgraph of 4 entities, (`A`, `B`, `C`, and `D`), with a set of some links connecting them as follows:
+
+```text
+           ┌─┐
+      ┌──► │B│ ───┐
+┌─┐   │    └─┘    │   ┌─┐
+│A│ ──┤           ├─► │D│
+└─┘   │    ┌─┐    │   └─┘
+      └──► │C│ ───┘
+           └─┘
+```
+
+Now, we could apply the same strategy as above, and create a "Path" entity type, and an associated "Has Element" link type.
+They could then use the _order_ of the link to maintain an ordered collection of the entities that make up the path.
```

Comment:
The order of what link? I'm not following this example (including the diagrams below), maybe easier to talk sync. This should be really clear as it's leaned on in the rationale

Reply:
The order of the `Has element` links that are displayed directly below

Reply:
https://github.com/blockprotocol/blockprotocol/pull/1020/commits/1d0294fa7f8e3ee53c19fd9524a5e50f841a51f4

https://github.com/blockprotocol/blockprotocol/pull/1020/commits/bb2b1fae04ba7a252be85444b5535741f2cbaccc

Reply:
[Make styling of diagrams consistent](https://github.com/blockprotocol/blockprotocol/pull/1020/commits/1d0294fa7f8e3ee53c19fd9524a5e50f841a51f4)


[Add clarification](https://github.com/blockprotocol/blockprotocol/pull/1020/commits/bb2b1fae04ba7a252be85444b5535741f2cbaccc)

Reply:
The clarification makes it much better, thank you